src/Resources/contao/library/Contao/System.php

```diff
@@ -587,7 +592,71 @@ public static function getImageSizes()
 			}
 		}
 
-		return static::$arrImageSizes;
+		$imageSizes = static::$arrImageSizes;
+
+		// Permission check
+		if (TL_MODE == 'BE' && $checkPermission === true)
+		{
+			$user = \BackendUser::getInstance();
+
+			// Limit only if the user is not an admin
+			if (!$user->isAdmin) 
+			{
+				$allowedSizes = deserialize($user->imageSizes, true);
+
+				if (empty($allowedSizes))
+				{
+					$imageSizes = [];
+				}
+				else
+				{
+					$filteredSizes = [];
+
+					foreach ($imageSizes as $group => $sizes) 
+					{
+						foreach ($sizes as $id => $size) 
+						{
+							// Dynamic sizes
+							if ($group == 'image_sizes') 
+							{
+								if (in_array($id, $allowedSizes)) 
+								{
+									$filteredSizes[$group][$id] = $size;
+								}
+
+								continue;
+							}
+
+							if (in_array($size, $allowedSizes)) 
+							{
+								$filteredSizes[$group][] = $size;
+							}
+						}
+					}
+
+					$imageSizes = $filteredSizes;
+				}
+			}
+		}
```

Comment:
As stated in https://github.com/contao/core-bundle/pull/319#issuecomment-126050836, all above here should be moved into a listener for the event being triggered below.